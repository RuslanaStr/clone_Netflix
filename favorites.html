<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
	<link rel="shortcut icon" href="img/favicon.ico" type="image/x-icon">
	<link rel="icon" href="img/favicon.ico" type="image/x-icon">
  <link rel="stylesheet" href="./css/favorite.css">
	<link rel="stylesheet" href="./css/style.css">
	<link rel="stylesheet" href="./css/media.css">
  <title>Сохраненные фильмы</title>
</head>
<body>
	<section class="wrapper">
		<header class="header">
			<div class="header__container" id="nav">
				<button class="header__burger-btn" id="burger">
					<span></span>
					<span></span>
					<span></span>
				</button>
				<a href="./index.html" class="header__logo">
					<img src="./img/netflix-logo.svg" alt="Logo">
				</a>
				<div class="header__menu menu">
					<nav class="menu__body">
						<ul class="menu__list">
							<li class="menu__item"><a href="./index.html" class="menu__link">Главная</a></li>
							<li class="menu__item"><a href="./films.html" class="menu__link">Фильмы</a></li>
							<li class="menu__item"><a href="./favorites.html" class="menu__link">Мой список</a></li>
						</ul>
					</nav>
					<div class="right__section">
						<div class="forms">
							<form class="search__container">
								<img class="search__icon" src="img/search.svg" alt="Search">
								<input type="text" class="search__input" placeholder="Поиск" />
							</form>
						</div>
						<div class="registration">
							<!-- <button class="form__btn">Вход</button> -->
							<img src="img/netflix-avatar.png" alt="Avatar" class="nav__avatar">
						</div>
					</div>
				</div>
			</div>
		</header>
		<div class="favorite__style">
			<h1 class="fav__title">Мой список</h1>
			<div id="saved__movies-container"></div>
			<div id="movie__template">
				<div class="movie">
					<h2 class="movie__title"></h2>
					<div class="movie__poster"></div>
				</div>
			</div>
		</div>
	</section>

	<footer class="footer">
		<div class="footer__container">
			<div class="footer__icons">
				<a href="https://www.facebook.com/netflix/"><img src="./img/icons/facebook.svg" alt="Facebook" class="footer__social"></a>
				<a href="https://www.instagram.com/Netflix/"><img src="./img/icons/instagram.svg" alt="Instagram" class="footer__social"></a>
				<a href="#"><img src="./img/icons/youtube.svg" alt="Youtube" class="footer__social"></a>
			</div>
			<div class="footer__link">
				<div class="column__left">
					<div class="link__column-one column">
						<a href="#" class="link__footer">Аудиоописание</a>
						<a href="#" class="link__footer">Связи с инвесторами</a>
						<a href="#" class="link__footer">Юридические сообщения</a>
					</div>
					<div class="link__column-two column">
						<a href="#" class="link__footer">Справочный центр</a>
						<a href="#" class="link__footer">Вакансии</a>
						<a href="#" class="link__footer">Настройка файлов cookie</a>
					</div>
				</div>
				<div class="column__right">
					<div class="link__column-three column">
						<a href="#" class="link__footer">Подарочные карты</a>
						<a href="#" class="link__footer">Условия использования</a>
						<a href="#" class="link__footer">Корпоративная информация</a>
					</div>
					<div class="link__column-four column">
						<a href="#" class="link__footer">Медиацентр</a>
						<a href="#" class="link__footer">Конфиденциальность</a>
						<a href="#" class="link__footer">Связаться с нами</a>
					</div>
				</div>
			</div>
		</div>
	</footer>

  <!-- Подключение скрипта, который отображает сохраненные фильмы -->
  <script type="module" src="public/JS/favorites.js"></script>
  <script type="module" src="./js/search.js"></script>
  <script type="module" src="./js/banner.js"></script>
  <script type="module" src="./js/header-film.js"></script>
  <script type="module" src="./js/movie-play.js"></script>
	<script type="module" src="./js/burger.js"></script>

<!-- Скрипт для работы видео плеера -->
	<script src="//kinoplayer.top/top.js"></script>

</body>
</html>
