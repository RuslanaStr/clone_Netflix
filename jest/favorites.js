"use strict";
// Функция для удаления фильма из localStorage
export function removeMovie(movieId) {
    // Получаем сохраненные фильмы из localStorage
    const savedMovies = JSON.parse(localStorage.getItem('savedMovies')) || [];
    // Фильтруем фильмы и оставляем только те, у которых id не совпадает с переданным movieId
    const updatedMovies = savedMovies.filter(movie => movie.id !== movieId);
    // Сохраняем обновленный массив в localStorage
    localStorage.setItem('savedMovies', JSON.stringify(updatedMovies));
    // Обновляем отображение сохраненных фильмов
    displaySavedMovies();
}
// Функция для отображения сохраненных фильмов на странице
export function displaySavedMovies() {
    // Получаем сохраненные фильмы из localStorage
    const savedMovies = JSON.parse(localStorage.getItem('savedMovies')) || [];
    // Контейнер, в котором будут показаны фильмы
    const savedMoviesContainer = document.getElementById('saved__movies-container');
    // Очищаем контейнер перед показом фильмов
    if(savedMoviesContainer){ //if

    savedMoviesContainer.innerHTML = '';
    }
    // Проходимся по каждому сохраненному фильму
    savedMovies.forEach(movie => {
        // Создаем элементы для отображения фильма
        const movieElement = document.createElement('div');
        movieElement.classList.add('movie');
        const posterElement = document.createElement('div');
        posterElement.classList.add('movie__poster');
        posterElement.innerHTML = `
      <img src="${movie.posterUrl}" alt="Movie Poster" class="favorite">`;
        // Переходим на страницу фильма при клике
        posterElement.onclick = () => {
            // window.open(`movie.html?id=${movie.id}&title=${encodeURIComponent(movie.title)}`, '_blank');
            window.location.href = `movie.html?id=${movie.id}&title=${encodeURIComponent(movie.title)}`;
        };
        // Создаем кнопку для удаления фильма
        const removeButton = document.createElement('button');
        removeButton.classList.add('remove__button');
        removeButton.textContent = 'Удалить';
        removeButton.addEventListener('click', () => {
            removeMovie(movie.id);
        });
        // Добавляем элементы фильма в контейнер
        // movieElement.appendChild(titleElement);
        movieElement.appendChild(posterElement);
        movieElement.appendChild(removeButton);
        if(movieElement && savedMoviesContainer){
        savedMoviesContainer.appendChild(movieElement);
    }
    });
}
// Вызываем функцию отображения сохраненных фильмов при загрузке страницы
window.addEventListener('DOMContentLoaded', displaySavedMovies);
