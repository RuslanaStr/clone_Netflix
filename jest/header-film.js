'use strict';

/*export const header = document.querySelector('.header');
window.addEventListener('scroll', () => {
  if (window.scrollY >= 100) {
    header.classList.add('header__black');
  } else {
    header.classList.remove('header__black');
  }
});*/

export function handleScroll() {
  const header = document.querySelector('.header');

  if (header) {
    window.addEventListener('scroll', () => {
      if (window.scrollY >= 100) {
        header.classList.add('header__black');
      } else {
        header.classList.remove('header__black');
      }
    });
  }
}

handleScroll(); 

