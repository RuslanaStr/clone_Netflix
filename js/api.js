'use strict';


// SWAGGER API
const API_KEY_KINOPOISK = '4d50104c-a87b-4ffd-883e-e4511dea4c8b';

// Картинка для баннера
const IMAGE_BANNER = 'https://kinopoiskapiunofficial.tech/api/v2.2/films/1044280/images?type=POSTER';

// Использовал для вытаскивания названия фильма по id для баннера
const API_URL_MOVIE_DETAILS = "https://kinopoiskapiunofficial.tech/api/v2.2/films/";
const MOVIE_ID = 1044280;


export { API_KEY_KINOPOISK, IMAGE_BANNER, API_URL_MOVIE_DETAILS, MOVIE_ID };
// banner.js

// ===========================================================================================

// TMBD API
// genre-list.js, search.js, modal.js, movie.js
const API_KEY_TMDB = '1b573b3ec7c63fed45643a781f334f2e';
const ID__FILM = 447365;
const URL_MOVIE = `https://api.themoviedb.org/3/movie/${ID__FILM}?api_key=${API_KEY_TMDB}&language=ru`;
const URL_TRAILERS = `https://api.themoviedb.org/3/movie/${ID__FILM}/videos?api_key=${API_KEY_TMDB}&language=uk`;
const IMAGE_BASE_URL = 'https://image.tmdb.org/t/p/';

export { API_KEY_TMDB, URL_MOVIE, URL_TRAILERS, IMAGE_BASE_URL, ID__FILM };