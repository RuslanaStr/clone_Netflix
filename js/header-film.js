'use strict';


const header = document.querySelector('.header');
window.addEventListener('scroll', () => {
  if (window.scrollY >= 100) {
    header.classList.add('header__black');
  } else {
    header.classList.remove('header__black');
  }
});