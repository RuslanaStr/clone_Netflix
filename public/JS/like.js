"use strict";
// Функция для сохранения фильма в localStorage
function saveMovie(movie) {
    // Получаем сохраненные фильмы из localStorage или создаем пустой массив
    const savedMovies = JSON.parse(localStorage.getItem('savedMovies')) || [];
    // Проверяем, не сохранен ли фильм уже
    const isSaved = savedMovies.some((savedMovie) => savedMovie.id === movie.id);
    if (!isSaved) {
        // Добавляем фильм в массив сохраненных фильмов
        savedMovies.push(movie);
        // Сохраняем обновленный массив в localStorage
        localStorage.setItem('savedMovies', JSON.stringify(savedMovies));
        // Изменяем стиль кнопки сердца
        const favoriteButton = document.getElementById('favorite__button');
        if (favoriteButton) {
            favoriteButton.classList.add('active');
        }
    }
    else {
        // Удаляем фильм из массива сохраненных фильмов
        const updatedMovies = savedMovies.filter((savedMovie) => savedMovie.id !== movie.id);
        localStorage.setItem('savedMovies', JSON.stringify(updatedMovies));
        // Изменяем стиль кнопки сердца
        const favoriteButton = document.getElementById('favorite__button');
        if (favoriteButton) {
            favoriteButton.classList.remove('active');
        }
    }
}
// Обработчик события клика на значок сердца
const favoriteButton = document.getElementById('favorite__button');
if (favoriteButton) {
    favoriteButton.addEventListener('click', () => {
        // Получаем информацию о фильме
        const movieTitleElement = document.getElementById('movie__title');
        const moviePoster = document.getElementById('poster__id');
        if (movieTitleElement && moviePoster) {
            const movie = {
                id: movieId || '',
                title: movieTitleElement.textContent || '',
                posterUrl: moviePoster.getAttribute('name') || '',
            };
            // Сохраняем или удаляем фильм
            saveMovie(movie);
        }
    });
}
// Получаем идентификатор фильма из параметров URL
const urlParams = new URLSearchParams(window.location.search);
const movieId = urlParams.get('id') || '';
