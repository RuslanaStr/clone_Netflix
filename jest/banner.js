'use strict';

import { ID__FILM, URL_MOVIE } from './api.js';

const bannerImageElement = document.querySelector('.banner');
const bannerTitleElement = document.querySelector('.banner__title');
const movieId = `${ID__FILM}`;

/**
 * Выполняет GET-запрос к API TMDb для получения данных о фильме.
 * @returns {Promise<Object>} Объект с данными о фильме.
 */
export async function getMovieDetails() {
  const url = `${URL_MOVIE}`;
  const resp = await fetch(url);
  const respData = await resp.json();
  return respData;
}

/**
 * Отображает баннер и название фильма.
 * @param {string} movieId - Идентификатор фильма.
 * @returns {Promise<void>}
 */
export async function showMovieBanner(movieId) {
  const bannerImageElement = document.querySelector('.banner');//добавлена выборка
  const bannerTitleElement = document.querySelector('.banner__title');

  // Получаем данные о фильме
  const movieDetails = await getMovieDetails(movieId);

  // Проверяем, есть ли изображение баннера у фильма
  if (movieDetails.backdrop_path) {
    const imageUrl = `https://image.tmdb.org/t/p/w1280${movieDetails.backdrop_path}`;
    const bannerElement = document.createElement('div');
    bannerElement.innerHTML = `
      <img src="${imageUrl}" alt="" class="banner__img">
    `;
    if (bannerImageElement && bannerElement) {    //выборка
    bannerImageElement.prepend(bannerElement);
  
  }
}
  // Устанавливаем название фильма в элемент баннера
  if (bannerTitleElement) {    //выборка
  bannerTitleElement.textContent = movieDetails.title;
}
}
// Вызываем функцию showMovieBanner с переданным идентификатором фильма
showMovieBanner(movieId);


// // Функция для выполнения GET-запроса к API TMDb для получения данных о фильме
// async function getMovieDetails() {
//   const url = `${URL_MOVIE}`;
//   const resp = await fetch(url);
//   const respData = await resp.json();
//   return respData;
// }

// // Функция для отображения баннера и названия фильма
// export async function showMovieBanner(movieId) {
//   const movieDetails = await getMovieDetails(movieId);

//   if (movieDetails.backdrop_path) {
//     const imageUrl = `https://image.tmdb.org/t/p/w1280${movieDetails.backdrop_path}`;
//     const bannerElement = document.createElement('div');
//     bannerElement.innerHTML = `
//       <img src="${imageUrl}" alt="" class="banner__img">
//     `;
//     bannerImageElement.prepend(bannerElement);
//   }

//   bannerTitleElement.textContent = movieDetails.title;
// }

// showMovieBanner(movieId);

