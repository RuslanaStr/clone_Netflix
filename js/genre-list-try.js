'use strict';

import { API_KEY_TMDB } from './api.js';

async function getGenre(genreId) {
  const url = `https://api.themoviedb.org/3/genre/movie/list?api_key=${API_KEY_TMDB}&language=ru`;
  const response = await fetch(url);
  const data = await response.json();

  if (response.ok) {
    const genre = data.genres.find(genre => genre.id === genreId);
    return genre;
  } else {
    throw new Error(data.status_message);
  }
}

async function getMoviesByGenre(genreId) {
  const url = `https://api.themoviedb.org/3/discover/movie?api_key=${API_KEY_TMDB}&with_genres=${genreId}&language=ru`;
  const response = await fetch(url);
  const data = await response.json();

  if (response.ok) {
    return data.results;
  } else {
    throw new Error(data.status_message);
  }
}

async function getPopularMovies() {
  const url = `https://api.themoviedb.org/3/movie/popular?api_key=${API_KEY_TMDB}&language=ru`;
  const response = await fetch(url);
  const data = await response.json();

  if (response.ok) {
    return data.results;
  } else {
    throw new Error(data.status_message);
  }
}

function showMoviesByGenre(genre) {
    const genreSections = document.querySelectorAll('.genre__section');

    genreSections.forEach(section => {
      if (section.classList.contains('genre__title')) {
        const moviesSection = section.nextElementSibling;
        const shouldDisplay = section.textContent === genre || genre === 'all';
        section.style.display = shouldDisplay ? 'block' : 'none';
        moviesSection.style.display = shouldDisplay ? 'block' : 'none';
      } else {
        const genreTitle = section.previousElementSibling;
        genreTitle.style.display = section.classList.contains(genre) || genre === 'all' ? 'block' : 'none';
        section.style.display = section.classList.contains(genre) || genre === 'all' ? 'block' : 'none';
      }
    });
}
  
async function showMoviesByGenres(genreIds) {
  const movieElement = document.querySelector('.genre__container');
  const genreContainer = document.createElement('div');
  genreContainer.classList.add('genres__movies');

  const popularMoviesElement = document.createElement('div');
  popularMoviesElement.classList.add('genre__section', 'popular', 'slider__section');
  const popularMoviesTitle = document.createElement('h2');
  popularMoviesTitle.classList.add('genre__title');
  popularMoviesTitle.textContent = 'Популярные фильмы';
  genreContainer.appendChild(popularMoviesTitle);

  const popularMovies = await getPopularMovies();
  for (const movie of popularMovies) {
    const movieElement = createMovieElement(movie);
    popularMoviesElement.appendChild(movieElement);
  }

  genreContainer.appendChild(popularMoviesElement);

  for (const genreId of genreIds) {
    const genreElement = document.createElement('div');
    const genre = await getGenre(genreId);

    genreElement.classList.add('genre__section', genre.id, 'slider__section');

    const genreTitle = document.createElement('h2');
    genreTitle.classList.add('genre__title');
    genreTitle.textContent = genre.name;

    genreContainer.appendChild(genreTitle);

    const movies = await getMoviesByGenre(genreId);
    for (const movie of movies) {
      const movieElement = createMovieElement(movie);
      genreElement.appendChild(movieElement);
    }

    genreContainer.appendChild(genreElement);
  }

  movieElement.innerHTML = '';
  movieElement.appendChild(genreContainer);

  const sliderElements = document.querySelectorAll('.slider__section');
  for (const sliderElement of sliderElements) {
    $(sliderElement).slick({
      slidesToShow: 6,
      slidesToScroll: 6,
      autoplay: false,
      autoplaySpeed: 3000,
      prevArrow: '<button type="button" class="slick-prev">Previous</button>',
      nextArrow: '<button type="button" class="slick-next">Next</button>',
      responsive: [
        {
            breakpoint: 1024,
            settings: {
                slidesToShow: 3,
                slidesToScroll: 3,
                infinite: true,
            }
        },
        {
            breakpoint: 600,
            settings: {
                slidesToShow: 2,
                slidesToScroll: 2
            }
        },
        {
            breakpoint: 480,
            settings: {
                slidesToShow: 2,
                slidesToScroll: 1
            }
        }
    ]
    });
  }
}

function createMovieElement(movie) {
  const sliderFilm = document.createElement('div');
  sliderFilm.classList.add('slider__film');

  const linkElement = document.createElement('a');
  // linkElement.setAttribute('target', '_blank');
  linkElement.classList.add('slider__poster-genre');

  const imageElement = document.createElement('img');
  imageElement.classList.add('slider__poster');
  imageElement.src = `https://image.tmdb.org/t/p/w500/${movie.poster_path}`;

  linkElement.appendChild(imageElement);
  sliderFilm.appendChild(linkElement);

  linkElement.addEventListener('click', () => {
    window.location.href = `movie.html?id=${movie.id}&title=${encodeURIComponent(movie.title)}`;
  });

  return sliderFilm;
}

document.addEventListener('DOMContentLoaded', () => {
  const selectGenre = document.getElementById('films__genre');
  selectGenre.addEventListener('change', function() {
    const selectedGenre = this.value;
    showMoviesByGenre(selectedGenre);
  });
    
  showMoviesByGenre('all');
});
  
const genreIds = [28, 12, 35, 18, 14];
showMoviesByGenres(genreIds)
  .catch(error => console.error(error));