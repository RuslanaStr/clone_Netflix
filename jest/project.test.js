/**
 * @jest-environment jsdom
 */

require = require('esm')(module);
import 'core-js/stable';
import 'regenerator-runtime/runtime';
import { JSDOM } from 'jsdom';
import { TextEncoder } from 'fast-text-encoding';
import fetch from 'jest-fetch-mock';

const dom = new JSDOM('<!doctype html><html><body></body></html>', {
  url: 'http://localhost',
  encoding: 'utf-8',
});
global.TextEncoder = TextEncoder;
global.fetch = fetch;


import { API_KEY_KINOPOISK, IMAGE_BANNER, API_URL_MOVIE_DETAILS, MOVIE_ID, API_KEY_TMDB } from './api.js';
import { showMovieBanner } from './banner.js';
import { handleScroll } from './header-film.js';
import { getGenre, getMoviesByGenre, getPopularMovies, showMoviesByGenre, showMoviesByGenres, createMovieElement } from './genre-list-try.js';
import { saveMovie } from './like.js';
import { removeMovie, displaySavedMovies } from './favorites.js';



jest.mock('./banner.js', () => {
  return {
    ...jest.requireActual('./banner.js'),
      getMovieDetails: jest.fn().mockResolvedValue({
      backdrop_path: '/path/to/image.jpg',
      title: 'Movie Title',
    }),
  };
});

describe('Banner Module', () => {
  test('API Key should be correct', () => {
    expect(API_KEY_KINOPOISK).toBe('4d50104c-a87b-4ffd-883e-e4511dea4c8b');
  });

  test('Banner image URL should be correct', () => {
    expect(IMAGE_BANNER).toBe('https://kinopoiskapiunofficial.tech/api/v2.2/films/1044280/images?type=POSTER');
  });

  test('API URL for movie details should be correct', () => {
    expect(API_URL_MOVIE_DETAILS).toBe('https://kinopoiskapiunofficial.tech/api/v2.2/films/');
  });

  test('Movie ID should be correct', () => {
    expect(MOVIE_ID).toBe(1044280);
  });
});

describe('ShowMovieBanner', () => {
    beforeEach(() => {
      document.body.innerHTML = `
        <div class="banner"></div>
        <h1 class="banner__title"></h1>
      `;
    });
  
    test('should display banner and movie title', async () => {
      const mockMovieDetails = {
        backdrop_path: '/path/to/image.jpg',
        title: 'Movie Title',
      };
  
      global.fetch = jest.fn().mockResolvedValue({
        ok: true,
        json: () => Promise.resolve(mockMovieDetails)
      });
  
      await showMovieBanner('123');
  
      const bannerImageElement = document.querySelector('.banner');
      const bannerTitleElement = document.querySelector('.banner__title');
  
      expect(bannerImageElement.innerHTML).toContain(`<img src="https://image.tmdb.org/t/p/w1280${mockMovieDetails.backdrop_path}" alt="" class="banner__img">`);
      expect(bannerTitleElement.textContent).toBe(mockMovieDetails.title);
    });
});


describe('Header Film Module', () => {
  let originalScrollY;
  let header;

  beforeEach(() => {
    originalScrollY = window.scrollY;
    document.body.innerHTML = '<div class="header"></div>';
    header = document.querySelector('.header');
    handleScroll();
  });

  afterEach(() => {
    window.scrollTo(0, originalScrollY);
  });

  test('should remove header__black class when scrolling below 100 pixels', () => {
    window.scrollTo(0, 50);
    window.dispatchEvent(new Event('scroll'));
    expect(header.classList.contains('header__black')).toBe(false);
  });
});

describe('getGenre', () => {
  test('should return genre by genreId', async () => {
    const genreId = 28;
    const mockResponse = {
      ok: true,
      json: jest.fn().mockResolvedValue({
        genres: [
          { id: 28, name: 'Action' },
          { id: 12, name: 'Adventure' },
        ],
      }),
    };
    global.fetch = jest.fn().mockResolvedValue(mockResponse);

    const genre = await getGenre(genreId);
    expect(genre).toEqual({ id: 28, name: 'Action' });
    expect(global.fetch).toHaveBeenCalledWith(
      `https://api.themoviedb.org/3/genre/movie/list?api_key=${API_KEY_TMDB}&language=ru`
    );
  });

  test('should throw an error if response is not OK', async () => {
    const genreId = 99;
    const mockResponse = {
      ok: false,
      json: jest.fn().mockResolvedValue({ status_message: 'Invalid genreId' }),
    };
    global.fetch = jest.fn().mockResolvedValue(mockResponse);

    await expect(getGenre(genreId)).rejects.toThrow('Invalid genreId');
    expect(global.fetch).toHaveBeenCalledWith(
      `https://api.themoviedb.org/3/genre/movie/list?api_key=${API_KEY_TMDB}&language=ru`
    );
  });
});

describe('getMoviesByGenre', () => {
  test('should return movies by genreId', async () => {
    const genreId = 28;
    const mockResponse = {
      ok: true,
      json: jest.fn().mockResolvedValue({
        results: [
          { id: 1, title: 'Movie 1' },
          { id: 2, title: 'Movie 2' },
        ],
      }),
    };
    global.fetch = jest.fn().mockResolvedValue(mockResponse);

    const movies = await getMoviesByGenre(genreId);
    expect(movies).toEqual([
      { id: 1, title: 'Movie 1' },
      { id: 2, title: 'Movie 2' },
    ]);
    expect(global.fetch).toHaveBeenCalledWith(
      `https://api.themoviedb.org/3/discover/movie?api_key=${API_KEY_TMDB}&with_genres=${genreId}&language=ru`
    );
});

  test('should throw an error if response is not OK', async () => {
    const genreId = 99;
    const mockResponse = {
      ok: false,
      json: jest.fn().mockResolvedValue({ status_message: 'Invalid genreId' }),
    };
    global.fetch = jest.fn().mockResolvedValue(mockResponse);

    await expect(getMoviesByGenre(genreId)).rejects.toThrow('Invalid genreId');
    expect(global.fetch).toHaveBeenCalledWith(
      `https://api.themoviedb.org/3/discover/movie?api_key=${API_KEY_TMDB}&with_genres=${genreId}&language=ru`
    );
  });
});

describe('getPopularMovies', () => {
  test('should return popular movies', async () => {
    const mockResponse = {
      ok: true,
      json: jest.fn().mockResolvedValue({
        results: [
          { id: 1, title: 'Movie 1' },
          { id: 2, title: 'Movie 2' },
        ],
      }),
    };
    global.fetch = jest.fn().mockResolvedValue(mockResponse);

    const movies = await getPopularMovies();
    expect(movies).toEqual([
      { id: 1, title: 'Movie 1' },
      { id: 2, title: 'Movie 2' },
    ]);
    expect(global.fetch).toHaveBeenCalledWith(
      `https://api.themoviedb.org/3/movie/popular?api_key=${API_KEY_TMDB}&language=ru`
    );
  });

  test('should throw an error if response is not OK', async () => {
    const mockResponse = {
      ok: false,
      json: jest.fn().mockResolvedValue({ status_message: 'Error' }),
    };
    global.fetch = jest.fn().mockResolvedValue(mockResponse);

    await expect(getPopularMovies()).rejects.toThrow('Error');
    expect(global.fetch).toHaveBeenCalledWith(
      `https://api.themoviedb.org/3/movie/popular?api_key=${API_KEY_TMDB}&language=ru`
    );
  });
});

describe('showMoviesByGenres', () => {
  beforeEach(() => {
    document.body.innerHTML = `
      <div class="genre__section genre__title" style="display: none;">
        Action
      </div>
      <div class="genre__section genre__movies" style="display: none;">
        <div class="movie movie-1">
          Movie 1
        </div>
        <div class="movie movie-2">
          Movie 2
        </div>
      </div>
    `;
  });


   test('should handle error', async () => {
    global.fetch = jest.fn().mockRejectedValueOnce(new Error('API error'));

    await expect(showMoviesByGenres([1])).rejects.toThrow('API error');

    const genreContainer = document.querySelector('.genre__container');
    if(genreContainer){
    expect(genreContainer.innerHTML).toBe('');
    }
  });
});


describe('createMovieElement', () => {
  test('should create movie element', () => {
    const movie = { id: 1, title: 'Movie 1', poster_path: 'poster1.jpg' };
    const movieElement = createMovieElement(movie);

    expect(movieElement).toMatchSnapshot();
  });
});

describe('showMoviesByGenre', () => {
  beforeEach(() => {
    document.body.innerHTML = `
      <div class="genre__section genre__title" style="display: none;">
        Action
      </div>
      <div class="genre__section genre__movies" style="display: none;">
        <div class="movie movie-1">
          Movie 1
        </div>
        <div class="movie movie-2">
          Movie 2
        </div>
      </div>
    `;
  });


  test('should hide movies for non-matching genre', () => {
    const genre = 'Drama';
    showMoviesByGenre(genre);

    const genreTitle = document.querySelector('.genre__title');
    const genreMovies = document.querySelector('.genre__movies');
    expect(genreTitle.style.display).toBe('none');
    expect(genreMovies.style.display).toBe('none');
  });

  test('should show movies for "all" genre', () => {
    const genre = 'all';
    showMoviesByGenre(genre);

    const genreTitle = document.querySelector('.genre__title');
    const genreMovies = document.querySelector('.genre__movies');
    expect(genreTitle.style.display).toBe('block');
    expect(genreMovies.style.display).toBe('block');
  });
});


const localStorageMock = (() => {
  let store = {};
  return {
    getItem: (key) => {
      return store[key];
    },
    setItem: (key, value) => {
      store[key] = value.toString();
    },
    removeItem: (key) => {
      delete store[key];
    },
    clear: () => {
      store = {};
    }
  };
})();

Object.defineProperty(window, 'localStorage', { value: localStorageMock });

describe('saveMovie', () => {
  beforeEach(() => {
    localStorage.clear(); 
  });

  test('should save a movie', () => {
    const movie = {
      id: '1',
      title: 'Test Movie',
      posterUrl: 'testposter.jpg'
    };
  
    const favoriteButton = document.createElement('button');
    favoriteButton.id = 'favorite__button';
    favoriteButton.classList.add('active');
  
    document.body.appendChild(favoriteButton);
  
    saveMovie(movie, favoriteButton);
  
    expect(localStorage.getItem('savedMovies')).toBeDefined();
    const savedMovies = JSON.parse(localStorage.getItem('savedMovies'));
    expect(savedMovies).toContainEqual(movie);
  
    expect(favoriteButton.classList.contains('active')).toBe(true);
  });
  
  test('should remove a movie', () => {
    const movie = {
      id: '1',
      title: 'Test Movie',
      posterUrl: 'testposter.jpg'
    };

   saveMovie(movie);

    saveMovie(movie);

    expect(localStorage.getItem('savedMovies')).toBeDefined();
    const savedMovies = JSON.parse(localStorage.getItem('savedMovies'));
    expect(savedMovies).toHaveLength(0);

    const favoriteButton = document.createElement('button');
    favoriteButton.id = 'favorite__button';
    saveMovie(movie);
    expect(favoriteButton.classList.contains('active')).toBe(false);
  });
});

describe('removeMovie', () => {
  beforeEach(() => {
    localStorage.clear(); 
  });

  test('should remove a movie from localStorage', () => {
    const savedMovies = [
      { id: '1', title: 'Movie 1', posterUrl: 'poster1.jpg' },
      { id: '2', title: 'Movie 2', posterUrl: 'poster2.jpg' },
      { id: '3', title: 'Movie 3', posterUrl: 'poster3.jpg' }
    ];
    localStorage.setItem('savedMovies', JSON.stringify(savedMovies));

    removeMovie('2');

    const updatedMovies = JSON.parse(localStorage.getItem('savedMovies'));
    expect(updatedMovies).toHaveLength(2);
    expect(updatedMovies).toEqual([
      { id: '1', title: 'Movie 1', posterUrl: 'poster1.jpg' },
      { id: '3', title: 'Movie 3', posterUrl: 'poster3.jpg' }
    ]);
  });
});

describe('displaySavedMovies', () => {
  test('should display saved movies on the page', () => {
    const savedMovies = [
      { id: '1', title: 'Movie 1', posterUrl: 'poster1.jpg' },
      { id: '2', title: 'Movie 2', posterUrl: 'poster2.jpg' }
    ];
    localStorage.setItem('savedMovies', JSON.stringify(savedMovies));

    document.body.innerHTML = `
      <div id="saved__movies-container"></div>
    `;

    displaySavedMovies();

    const savedMoviesContainer = document.getElementById('saved__movies-container');
    if(savedMoviesContainer){
    expect(savedMoviesContainer.children).toHaveLength(2);
    expect(savedMoviesContainer.children[0].querySelector('.movie__poster img').getAttribute('src')).toBe('poster1.jpg');
    expect(savedMoviesContainer.children[1].querySelector('.movie__poster img').getAttribute('src')).toBe('poster2.jpg');
  }
  });
});














