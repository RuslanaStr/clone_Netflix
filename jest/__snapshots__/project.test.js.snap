// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMovieElement should create movie element 1`] = `
<div
  class="slider__film"
>
  <a
    class="slider__poster-genre"
    target="_blank"
  >
    <img
      class="slider__poster"
      src="https://image.tmdb.org/t/p/w500/poster1.jpg"
    />
  </a>
</div>
`;
